components/TextFields/MDCTextInput.h

```diff
@@ -0,0 +1,191 @@
+   the field is being edited. The character count is below text. The Material Design guideslines
+   call this 'Floating inline labels.'
+   https://material.io/guidelines/components/text-fields.html#text-fields-labels
+   */
+  MDCTextInputPresentationStyleFloatingPlaceholder,
+
+  /** The placeholder is laid out inline and the character count is also inline to the right. */
+  MDCTextInputPresentationStyleFullWidth,
+};
+
+/** Delegate for MDCTextInput size changes. */
+@protocol MDCTextInputLayoutDelegate <NSObject>
+
+@optional
+/**
+ Notifies the delegate that the textInput's content size changed, requiring the size provided.
+
+ @param textInput The text input for which the content size changed.
+ @param size      The size required by the text input to fit all of its content.
+ */
+- (void)textInput:(_Nonnull id<MDCTextInput>)textInput contentSizeChanged:(CGSize)size;
+
+@end
+
+/** Behavior exhibited by Material Design themed text inputs. */
+@protocol MDCTextInput <NSObject>
+
+/** A Boolean value indicating whether the text field is currently in edit mode. */
+@property(nonatomic, readonly, getter=isEditing) BOOL editing;
+
+/**
+ The text to be displayed when no text has been entered.
+
+ If presentationStyle == MDCTextInputPresentationStyleFloatingPlaceholder this text will also float
+ above the input when text has been entered. The Material Design guidelines call this 'Hint text'.
+ https://material.io/guidelines/components/text-fields.html#text-fields-input
+ */
+@property(nonatomic, nullable, copy) IBInspectable NSString *placeholder;
+
+/**
+ The color applied to the placeholder when inline (not floating.)
+
+ Default is black with Material Design hint text opacity.
+ */
+@property(nonatomic, nullable, strong) UIColor *inlinePlaceholderColor UI_APPEARANCE_SELECTOR;
+
+/**
+ The color applied to the placeholder when floating. However, when in error state, it will be
+ colored with the error color.
+
+ Default is black with Material Design hint text opacity.
+ */
+@property(nonatomic, nullable, strong) UIColor *floatingPlaceholderColor UI_APPEARANCE_SELECTOR;
+
+/** The text displayed in the text input. */
+@property(nonatomic, nullable, copy) NSString *text;
+
+/**
+ The text displayed under input rectangle, when in error state, that explains the nature of the
+ error.
+
+ The value of this property controls the state of the text input. When errorText != nil, the text
+ input is in an error state:
+  - The error text appears in a label below the input rectangle's underline with the errorColor as
+  text color.
+  - The input rectangle's underline is colored with the errorColor.
+ */
+@property(nonatomic, nullable, copy, readonly) IBInspectable NSString *errorText;
+
+/**
+ The color used to denote error state in the underline and the errorText's label.
+
+ Default is red.
+ */
+@property(nonatomic, nullable, copy) UIColor *errorColor UI_APPEARANCE_SELECTOR;
```

Comment:
question: when set this doesn't do anything except when the textInput is in a specific state?
when an internal validator (character count) or external validator (setErrorText:) has decided that it is an error?

Now that we have properties for the line color and others. Could it be the responsibility of the validator to just set those properties directly?